apps/expo/src/components/ui/keypad/Keypad.tsx

```diff
@@ -0,0 +1,177 @@
+import React, { useState } from 'react'
+import { Dimensions, Text, TouchableOpacity, View } from 'react-native'
+
+import { AntDesign, Ionicons } from '@expo/vector-icons'
+
+interface NumericKeypadProps {
+  setNumber: React.Dispatch<React.SetStateAction<any>>
+  number: string
+  keypadVisible: boolean
+  setKeypadVisible: React.Dispatch<React.SetStateAction<boolean>>
+}
+
+export default function Keypad({
+  setNumber,
+  number,
+  keypadVisible,
+  setKeypadVisible,
+}: NumericKeypadProps) {
+  const handleKeyPress = (value: string) => {
+    /* istanbul ignore next */
+    if (value === 'backspace') {
+      setNumber(number.slice(0, -1)) // remove the last digit
+    } else if (value === '.') {
+      if (!number.includes('.')) {
+        // only add a '.' if there is not one already
+        setNumber(number + value)
+      }
+    } else if (number === '0') {
+      // if the only thing in the string is a 0, just replace the 0 with the digit
+      setNumber(value)
+    } else if (
+      (number.includes('.') && number.indexOf('.') === number.length - 3) ||
+      Number(number + value) > 9999.99
+    ) {
+      // do not allow further input if they want more than 2 decimal points of precision
+      // or if they want to enter a value of 10,000 or more
+      // preventing any buffer overflow attempts
+    } else if (Number(value) >= 0 && Number(value) <= 9) {
+      // else just concatenate as normal
+      setNumber(number + value)
+    }
+  }
+
+  const { height } = Dimensions.get('window')
+
+  return !keypadVisible ? (
+    <View testID='invisible-test' />
+  ) : (
+    <View className='flex bg-slate-200 p-1' style={{ height: height * 0.35 }}>
+      <View className='flex-1 flex-row justify-between'>
```

Comment:
I think all of these could be refactored with an array and a `map()` call. Would make this super clean.

Reply:
Yeah i originally had it like this but it made unit testing a pAIN ngl so i just reverted back